In [3]:
import requests
from bs4 import BeautifulSoup

## enumerate() 

In [1]:
names = ['Lee', 'Kim', 'Choi']
list(enumerate(names))

[(0, 'Lee'), (1, 'Kim'), (2, 'Choi')]

In [2]:
dict(enumerate(names))

{0: 'Lee', 1: 'Kim', 2: 'Choi'}

### for문만으로 묶기

In [10]:
l = []
d = {}

for i in range(len(names)):
    l.append((i, names[i]))
    d[i] = names[i]
    
print(l)
print(d)

[(0, 'Lee'), (1, 'Kim'), (2, 'Choi')]
{0: 'Lee', 1: 'Kim', 2: 'Choi'}


### enumerate() 사용하기

In [11]:
l = []
d = {}

for i, name in enumerate(names):
    l.append((i, name))
    d[i] = names[i]
    
print(l)
print(d)

[(0, 'Lee'), (1, 'Kim'), (2, 'Choi')]
{0: 'Lee', 1: 'Kim', 2: 'Choi'}


### 예제 - 인덱스 번호를 붙여보자

In [19]:
res = requests.get('https://movie.naver.com/movie/bi/mi/pointWriteFormList.naver?code=45290&type=after&isActualPointWriteExecute=false&isMileageSubscriptionAlready=false&isMileageSubscriptionReject=false&page=4930')
soup = BeautifulSoup(res.content, 'html.parser')

id_list = []
id_pre = '_filtered_ment_'
for i in range(10):
    id_list.append(id_pre+str(i))

mydata = []
for id in id_list:
    mydata.append(soup.find('span', {'id':id}).get_text().strip())

for index, item in enumerate(mydata):
    print(index, item)

0 영화는 볼만했는데 구미강동점  캬라멜 팝콘 너무 맛없네요 캬라멜을 쥐똥만큼 아껴 넣어서는 4500원이나 받아먹고 너무하네요 !!
1 억지 영화...노잼 이런게 천만 ㅋㅋ
2 이게 재밌다구요?물론 초중반까지는 재밌었어요후반에 진짜 어휴
3 본인이 SF잘모른다 그러면 보지마세요 더럽게 재미없습니다 전 스물한살이고 SF이런거 흥미없었던 지라 너무 재미없었습니다
4 이거 결말 아빠가 유령임 이거 결말 아빠가 유령임이거 결말 아빠가 유령임 이거 결말 아빠가 유령임이거 결말 아빠가 유령임 이거 결말 아빠가 유령임이거 결말 아빠가 유령임 이거 결말 아빠가 유령임이거 결말 아빠가 유령...
5 평점 1점도 아까운 저렴 신파 영화입니다. 보다가 도저히 볼수 없어 중간에 나왔어요.
6 TV에서 콘스탄틴, 소림축구, 쿵푸허슬 나오면 내용알지만 다시 보게된다이영화는 영화관에서보구 그후 티비에서 나와도 거른다 솔직히 이영화가 재밌냐?취향 때문이 그럴수도 있지만 이기 영화냐?!!!???? 다큐멘터리나?
7 열라 지루하네... 여자분들은 보지마세요. 하품하다 나옵니다.
8 별로 재미없습니다...
9 연기도못하고 영화최악이다 졸려죽는줄알았네


### split() 함수를 이용한 전처리

In [20]:
txt = '(월세) - 명지대 앞 서브웨이 뒷건물 원룸 500/50 [5]'

In [21]:
txt.split('[')[0]

'(월세) - 명지대 앞 서브웨이 뒷건물 원룸 500/50 '

In [22]:
txt.split('[')[0].split('-')

['(월세) ', ' 명지대 앞 서브웨이 뒷건물 원룸 500/50 ']

In [23]:
txt.split('[')[0].split('-')[1]

' 명지대 앞 서브웨이 뒷건물 원룸 500/50 '

In [25]:
txt.split('[')[0].split('-')[1].strip()

'명지대 앞 서브웨이 뒷건물 원룸 500/50'

# 정규표현식

In [26]:
data = """
park 800905-1049118
kim  700905-1059119
"""

In [27]:
result = []
for line in data.split('\n'):
    word_result = []
    for word in line.split(' '):
        if len(word) == 14 and word[:6].isdigit() and word[7:].isdigit():
            word = word[:6] + '-' + '*******'
        word_result.append(word)
    result.append(" ".join(word_result))
print('\n'.join(result))


park 800905-*******
kim  700905-*******



In [28]:
import re

pat = re.compile('(\d{6})[-]\d{7}') # 6자리 숫자 - 7자리 숫자 (정규표현식) 패턴화한 결과
print(pat.sub('\g<1>-*******', data))


park 800905-*******
kim  700905-*******



### 메타문자 
 - .^$*+?{}[]\|()
 - [abc] = a, b, c 중 한개의 문자를 포함
     * 'a' = 매치 / before = 매치 / dude = 매치X

# 1. Dot .
        - \n 제외한 모든문자 한개
        - ex) D.A = D + 모든문자(한개) + A

In [29]:
pattern = re.compile('D.A')

In [30]:
pattern.search('DDA')

<re.Match object; span=(0, 3), match='DDA'>

In [31]:
pattern.search('D1A')

<re.Match object; span=(0, 3), match='D1A'>

In [32]:
pattern.search('D00A')

In [34]:
pattern.search('DA')

In [36]:
pattern.search('d0A')

In [37]:
pattern.search('d0A D1A 0111')

<re.Match object; span=(4, 7), match='D1A'>

## real Dot . 를 찾으려면?

In [38]:
pattern = re.compile('D\.A')

In [39]:
pattern.search('D.A')

<re.Match object; span=(0, 3), match='D.A'>

In [40]:
pattern.search("DDA")

In [41]:
pattern = re.compile('D[.]A')

## 찾고 바꾸기

In [42]:
string = 'DDA D1A DDA DA'

In [43]:
re.sub('D.A', 'Dave', string)

'Dave Dave Dave DA'

## 2. 반복 ? , * +
    - ? = 앞 문자가 0번 / 1번 표시
    - * = 앞 문자가 0번 이상
    - + = 앞 문자가 1번 이상

### ?

In [44]:
pattern = re.compile('D?A')

In [45]:
pattern.search("A")

<re.Match object; span=(0, 1), match='A'>

In [46]:
pattern.search('DA')

<re.Match object; span=(0, 2), match='DA'>

In [47]:
pattern.search('DDDDDDA')

<re.Match object; span=(5, 7), match='DA'>

### *

In [48]:
pattern = re.compile('D*A')

In [49]:
pattern.search('A')

<re.Match object; span=(0, 1), match='A'>

In [50]:
pattern.search('DA')

<re.Match object; span=(0, 2), match='DA'>

In [53]:
pattern.search('DDDDDDDDDDDDDDDDDDDDDDDDDDDDA')

<re.Match object; span=(0, 29), match='DDDDDDDDDDDDDDDDDDDDDDDDDDDDA'>

In [54]:
pattern = re.compile('AD*A')

In [55]:
pattern.search("ADA")

<re.Match object; span=(0, 3), match='ADA'>

In [57]:
pattern.search("ADDDDDDDDDDDDDDDDDDA")

<re.Match object; span=(0, 20), match='ADDDDDDDDDDDDDDDDDDA'>

### +

In [58]:
pattern = re.compile('D+A')

In [59]:
pattern.search('A')

In [60]:
pattern.search('DA')

<re.Match object; span=(0, 2), match='DA'>

In [63]:
pattern.search("DDDDDDDDDDDDDDDDDDDDDDDDDDDDA")

<re.Match object; span=(0, 29), match='DDDDDDDDDDDDDDDDDDDDDDDDDDDDA'>

### {n}, {m, n}
    - {n} : 앞문자 n번 반복
    - {m, n} : 앞 문자가 m번 반복 ~ n 번 반복 (ex. {3, 5} = 3번, 4번, 5번 반복)

### {n} - 잘 안쓰임

In [64]:
pattern = re.compile('AD{2}A')

In [65]:
pattern.search('ADA')

In [66]:
pattern.search('ADDA')

<re.Match object; span=(0, 4), match='ADDA'>

In [67]:
pattern.search('ADDDA')

### {m, n} 

In [71]:
pattern = re.compile('AD{2,6}A')

In [72]:
pattern.search('ADDA')

<re.Match object; span=(0, 4), match='ADDA'>

In [73]:
pattern.search('ADDDA')

<re.Match object; span=(0, 5), match='ADDDA'>

In [75]:
pattern.search('ADDDDDDA')

<re.Match object; span=(0, 8), match='ADDDDDDA'>

## 3. [] : 안에 들어가는 문자가 들어있는 패턴
    - [abc] : a, b, c 중 하나가 들어있는 패턴

In [76]:
pattern = re.compile('[abcdefgABCDEFG]')

In [77]:
pattern.search('a1234')

<re.Match object; span=(0, 1), match='a'>

In [78]:
pattern.search('z1234')

### 하이픈(-)을 이용하면 알파벳 전체 나타낼 수 있음
    - [a-c] : a, b, c 중 하나가 들어있는 패턴

In [79]:
pattern = re.compile('[a-z]')

In [80]:
pattern.search('k1234')

<re.Match object; span=(0, 1), match='k'>

In [81]:
pattern.search('Z1234')

### [a-zA-Z] : 대소문자 모두 포함한 알파벳 전체

In [82]:
pattern = re.compile('[a-zA-Z]')

In [83]:
pattern.search('Z1234')

<re.Match object; span=(0, 1), match='Z'>

### [a-zA-Z0-9]로 표기하면 대소문자를 모두 포함해서 알파벳 전체와 함께 숫자전체

In [84]:
pattern = re.compile('[a-zA-Z0-9]')

In [85]:
pattern.search('1234---')

<re.Match object; span=(0, 1), match='1'>

In [87]:
pattern.search('--------------_!@#!@$#%!#$----')

### [ ] 괄호 안에서 [ 바로 뒤에 ^을 쓰면 그 뒤에 오는 문자가 아닌 패턴 찾기

In [88]:
pattern = re.compile('[^a-zA-Z0-9]')

In [89]:
pattern.search('--------------_!@#!@$#%!#$----')

<re.Match object; span=(0, 1), match='-'>

In [90]:
pattern = re.compile('[^\t\n\r\f\v]')

In [91]:
pattern.search('-')

<re.Match object; span=(0, 1), match='-'>

In [92]:
pattern.search(' ')

<re.Match object; span=(0, 1), match=' '>

## 한글만? -> [가-힣]

In [93]:
pattern = re.compile('[가-힣]')

In [94]:
pattern.search('안')

<re.Match object; span=(0, 1), match='안'>

# 4. 정규표현식 라이브러리 함수 사용법

## match, search 함수
    - match : 문자열 처음부터 정규식과 매칭되는 패턴 찾아서 리턴 / 전체
    - search : 문자열 전체를 검색해서 정규식과 매칭되는 패턴 찾아서 리턴 / 부분

In [95]:
pattern = re.compile('[a-z]+')

In [100]:
matched = pattern.match('Dave') # 처음 시작이 대문자라서 리턴 X

In [97]:
searched = pattern.search('Dave')

In [98]:
print(matched)

None


In [99]:
print(searched)

<re.Match object; span=(1, 4), match='ave'>


## findall 함수 : 정규표현식과 매칭되는 모든 문자열을 리스트 객체로 리턴

In [101]:
pattern = re.compile('[a-z]+')
findalled = pattern.findall('Game of Lite in Python')

In [102]:
print(findalled)

['ame', 'of', 'ite', 'in', 'ython']


In [106]:
pattern2 = re.compile('[A-Za-z]+') # 뒤에 왜 +가 붙을까

In [104]:
findalled2 = pattern2.findall('Game of Lite in Python')

In [105]:
print(findalled2)

['Game', 'of', 'Lite', 'in', 'Python']


## findall 함수를 사용해서 정규표현식에 해당되는 문자열 확인

In [108]:
pattern = re.compile('[a-z]+')
findalled = pattern.findall('GAME')

if len(findalled)>0:
    print('정규표현식에 맞는 문자열 존재')
else:
    print('정규표현식에 맞는 문자열 존재 안함')

정규표현식에 맞는 문자열 존재 안함


## split 함수 : 찾은 정규표현식 패턴 문자열을 기준으로 문자열 분리

In [109]:
pattern2 = re.compile(':')
splitted = pattern2.split('python:java')

In [110]:
print(splitted)

['python', 'java']


---

In [518]:
import re

###    1. vs 기준으로 나누기

In [527]:
name_s =  'Spider man vs Super man'
name_i = 'Iron man vs. Ultron'

In [528]:
pattern = re.compile(' vs. ')
print(pattern.split(name_s))
print(pattern.split(name_i))

['Spider man vs Super man']
['Iron man', 'Ultron']


###    2. 계좌에서 숫자만 추출

In [521]:
kb = 'KB 231401-04-296081'
nh = 'NH 301-0173-716502'

In [522]:
pattern = re.compile('[^0-9]')
new_kb = re.sub(pattern,'', kb)
new_nh = re.sub(pattern, '', nh)

In [523]:
print(new_kb)
print(new_nh)

23140104296081
3010173716502


###    1. 학번에서 입학년도만 **으로 변경하기
        * 60191234 -> 60**1234
        * 60153456 -> 60**3456

In [390]:
st_num = '60191095'

In [421]:
pattern = re.compile(r'(\d{2})(\d{2})(\d{4})')
pattern.sub('\g<1>**\g<3>',st_num)

'60**1095'

###    2. 비번 인증 만들기 
        * 대소문자, 숫자, 특수기호(#, @, &) 각각 하나 이상 포함
        * 전체 8글자 이상, 13자리 미만
            - abc12345 : X 
            - aA#1 : X
            - aA#12345 : O
            - abcdefgABC#@&123 : X

In [515]:
password = ['abc12345', 'aA#1', 'aA#12345', 'bcdefgABC#@&123', 
           'abcdefghijk', '12345', 'Abcd@efg123']

In [516]:
pattern = re.compile("^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@#&])[A-Za-z\d@#&]{8,13}$")

    -> "^(?=.*[a-z])(?=.*[A-Z])(?=.*\d)(?=.*[@#&])[A-Za-z\d@#&]{8,13}$"

In [517]:
for i in password:
    a = pattern.search(i)
    if a is not None:
        print(i, '비밀번호 조건 만족')
    else:
        print(i, '비밀번호 조건 불만족')        

abc12345 비밀번호 조건 불만족
aA#1 비밀번호 조건 불만족
aA#12345 비밀번호 조건 만족
bcdefgABC#@&123 비밀번호 조건 불만족
abcdefghijk 비밀번호 조건 불만족
12345 비밀번호 조건 불만족
Abcd@efg123 비밀번호 조건 만족



###    3. Mr은 남성, Miss는 미혼여성, Mrs는 기혼여성 나머지는 기타로 분류
        * 성별 컬럼 내용 무시 
        * 분류결과를 별도의 컬럼 Title에 저장
        * 위에서 만든 컬럼의 종류에 따른 생존율 구하기 
            - 생존률 = 생존자수 / (생존자수 + 사망자수) : Title grouping

In [163]:
import pandas as pd
import numpy as np

In [532]:
df = pd.read_csv('train.csv')

In [533]:
df.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [534]:
pattern_mr = re.compile('Mr[^s]')
pattern_mrs = re.compile('Mrs')
pattern_miss = re.compile('Miss')

In [535]:
l = []
for i in range(len(df)):
    if pattern_mr.search(df['Name'][i]) is not None:
        l.append('Mr')
    elif pattern_mrs.search(df['Name'][i]) is not None:
        l.append('Mrs')
    elif pattern_miss.search(df['Name'][i]) is not None:
        l.append('Miss')
    else:
        l.append('기타')

In [536]:
df['Title'] = l

In [537]:
def s_rate(x):
    print(x['Title'].unique(), (x['Survived'].value_counts()[1]/len(x)), '\n')

In [538]:
df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked Title  
0      0         A/5 21171   7.2500   NaN        S    Mr  
1      0          PC 17599  71.2833   C85        C   Mrs  
2      0  STON/O2. 3101282   7.9250   NaN        S  Miss

In [539]:
df.groupby(['Title']).apply(s_rate)

['Miss'] 0.7 

['Mr'] 0.1583011583011583 

['Mrs'] 0.7906976744186046 

['기타'] 0.5 



Empty DataFrame
Columns: []
Index: []

    - 미혼 여성의 생존률 = 0.7
    - 남성의 생존률 = 0.15
    - 기혼 여성의 생존률 = 0.79
    - 기타 생존률 = 0.5